[JIT] Add aliasing checks in SchemaInfo with associated tests

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81352
* #81007
* #81002
* #81000
* __->__ #80984
* #80972
* #80918

- Created may_alias method in SchemaInfo to update the implementation of FunctionSchema::may_alias for aliasing cases due to inputs aliasing. 
- Created output_alias_map_ internal variable to check cases where outputs might alias due to inputs aliasing. This variable is updated in generateAliasMap().
- Added tests for various may_alias special cases (input - input, input - output, output - output) due to inputs aliasing causing other arguments to also alias.

Commit:
[JIT] Add aliasing checks in SchemaInfo with associated tests

[ghstack-poisoned]

Commit:
Update on "[JIT] Add aliasing checks in SchemaInfo with associated tests"


- Created areAliasing method in SchemaInfo to update the implementation for the check_additional boolean parameter implementation. This implementation will also return true in cases where arguments alias due to inputs aliasing. 
- Created output_alias_map_ internal variable to check cases where outputs might alias due to inputs aliasing.
- Added tests for various areAliasing special cases (input - input, input - output, output - output) due to inputs aliasing causing other arguments to also alias.

[ghstack-poisoned]

Commit:
Update on "[JIT] Add aliasing checks in SchemaInfo with associated tests"


- Created areAliasing method in SchemaInfo to update the implementation for the check_additional boolean parameter implementation. This implementation will also return true in cases where arguments alias due to inputs aliasing. 
- Created output_alias_map_ internal variable to check cases where outputs might alias due to inputs aliasing.
- Added tests for various areAliasing special cases (input - input, input - output, output - output) due to inputs aliasing causing other arguments to also alias.

[ghstack-poisoned]

davidberard98 reviewed (commented):
Looks good, I just recommend a slight refactor of may_alias (see the comment)

Commit:
Update on "[JIT] Add aliasing checks in SchemaInfo with associated tests"


- Created may_alias method in SchemaInfo to update the implementation of FunctionSchema::may_alias for aliasing cases due to inputs aliasing. 
- Created output_alias_map_ internal variable to check cases where outputs might alias due to inputs aliasing. This variable is updated in generateAliasMap().
- Added tests for various may_alias special cases (input - input, input - output, output - output) due to inputs aliasing causing other arguments to also alias.

[ghstack-poisoned]

Gamrix commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -58,10 +58,43 @@ bool SchemaInfo::is_mutable(c10::string_view name) {
+  } else if (
+      lhs.type == c10::SchemaArgType::output &&
+      rhs.type == c10::SchemaArgType::output) {
+    for (size_t lhs_alias_input : output_alias_map_[lhs.index]) {
+      for (size_t rhs_alias_input : output_alias_map_[rhs.index]) {
+        if (lhs_alias_input == rhs_alias_input) {
+          return true;
+        }
+      }
```
As these are maps, we should be able to just do `output_alias_map_[rhs.index].count(lhs_alias_input)` and eliminate the double for loop.

Commit:
Update on "[JIT] Add aliasing checks in SchemaInfo with associated tests"


- Created may_alias method in SchemaInfo to update the implementation of FunctionSchema::may_alias for aliasing cases due to inputs aliasing. 
- Created output_alias_map_ internal variable to check cases where outputs might alias due to inputs aliasing. This variable is updated in generateAliasMap().
- Added tests for various may_alias special cases (input - input, input - output, output - output) due to inputs aliasing causing other arguments to also alias.

[ghstack-poisoned]

goldenxuett commented:
@pytorchbot merge
